src/charm.py

```diff
@@ -4,110 +4,123 @@
 )
-from mysql_router_helpers import MySQLRouter
-from relations.database_provides import DatabaseProvidesRelation
-from relations.database_requires import DatabaseRequiresRelation
-from relations.tls import MySQLRouterTLS
 
 logger = logging.getLogger(__name__)
 
 
-class MySQLRouterOperatorCharm(CharmBase):
-    """Operator charm for MySQLRouter."""
+class MySQLRouterOperatorCharm(ops.charm.CharmBase):
+    """Operator charm for MySQL Router."""
 
-    def __init__(self, *args):
+    def __init__(self, *args) -> None:
         super().__init__(*args)
 
-        self.framework.observe(self.on.install, self._on_install)
-        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.database_requires = relations.database_requires.Relation(
+            data_interfaces.DatabaseRequires(
+                self,
+                relation_name=DATABASE_REQUIRES_RELATION,
+                # HACK: mysqlrouter needs a user, but not a database
+                # Use the DatabaseRequires interface to get a user; disregard the database
+                database_name="_unused_mysqlrouter_database",
+                extra_user_roles="mysqlrouter",
+            )
+        )
         self.framework.observe(
-            getattr(self.on, "mysql_router_pebble_ready"), self._on_mysql_router_pebble_ready
+            self.database_requires.interface.on.database_created,
+            self._reconcile_database_relations,
+        )
+        self.framework.observe(
+            self.on[DATABASE_REQUIRES_RELATION].relation_broken,
+            self._reconcile_database_relations,
         )
-        self.framework.observe(self.on[PEER].relation_changed, self._on_peer_relation_changed)
-        self.framework.observe(self.on.update_status, self._on_update_status)
-
-        self.database_provides = DatabaseProvidesRelation(self)
-        self.database_requires = DatabaseRequiresRelation(self)
-        self.tls = MySQLRouterTLS(self)
-
-    # =======================
-    #  Properties
-    # =======================
-
-    @property
-    def peers(self) -> Optional[Relation]:
-        """Fetch the peer relation."""
-        return self.model.get_relation(PEER)
 
-    @property
-    def app_peer_data(self):
-        """Application peer data object."""
-        if not self.peers:
-            return {}
+        self.database_provides = relations.database_provides.Relation(
+            data_interfaces.DatabaseProvides(self, relation_name=DATABASE_PROVIDES_RELATION)
+        )
+        self.framework.observe(
+            self.database_provides.interface.on.database_requested,
+            self._reconcile_database_relations,
+        )
+        self.framework.observe(
+            self.on[DATABASE_PROVIDES_RELATION].relation_broken, self._reconcile_database_relations
+        )
 
-        return self.peers.data[self.app]
+        self.framework.observe(
+            getattr(self.on, "mysql_router_pebble_ready"), self._on_mysql_router_pebble_ready
```

Comment:
sorry, this is completely optional, but since these are objects defined at runtime only, noisy linters don't know about their attributes and doing so shut them up. I like to keep the linter noisy because most of the times you wan't to know